### DISCUSSION TITLE
Split 4bit color theme and 256color themes?

### DISCUSSION THREAD

**Discussion by jinek** [Consolonia/Consolonia#166] [Ideas]
@tomlm Since we have introduced 256 colors, do we assume 16 colors limitation is not needed anymore? Can we hit a case (some terminal for instance) which does not support 256 colors?
I'm thinking whether we should maintain 4bit mode?
There is also a separate task to support monochrome devices, like Beepberry
![image](https://github.com/user-attachments/assets/b085b636-433a-4391-be52-dbd487c868af)

**Comment by tomlm** [comment]
Seems like it's time for us to introduce a PlatformSettings option.
```c#
enum ColorMode { 256Bit, 3Bit, 4Bit, Monochrome };

public class ConsoloniaPlatformSettings
{
    public ColorMode ColorMode {get;set;}
}
```

And then at render time we could map colors to color map based on settings. 
* For 256Bit=> Output full 256 ESC sequence like today
* For 4Bit=> Output using 4 bit ESC Sequence (16 colors)
* For 3Bit=> Output using 3 bit ESC Sequence (8 colors)
* For Mono => No color escaping at all, giving us pure text.
Something like that. 

We could map the RGB colors to underlying color easy enough, and standard colors like Colors.Red => esc(red)

**Reply by tomlm** [reply]
There would be a perf enhancement for Monochrome mode, as no escapes for colors.

**Reply by tomlm** [reply]
As I think about it more, we would probably need to do this via themes support...I don't think just arbitrarily mapping colors is going look great.

**Reply by tomlm** [reply]
Interesting. As I think about monochrome, I think we would have to leverage Inverted black and white for things like selection in a listbox.

**Reply by jinek** [reply]
True, I'm thinking about the same. Color scaling/interpolating will not work in general. Monochrome is a good edge case to validate against.

Here are my thoughts:
1) I've almost finished converting to ControlTheme
2) I see that we already have split to several themes (TurboVision/Dark/Black, Fluent, Material), it's already done, I did not see that, thus this topic is already complete.
3) What I don't understand is how we are supposed to deal with different color modes when drawing an image? Also similar question about templates. What if someone tries to draw red border in monochrome console? Should it be legit, should we interpolate?
4) Speaking about monochrome - it for sure will not be enough to just even change the color palette. But in some cases will be necessary to override some properties (like background, selected background, borders etc etc). And in some tough cases will be necessary to override the template itself (still may be enough to change TurboVision template to fit everything dependending on properties only).

**Comment by jinek** [comment]
@tomlm 
Ah ok, I somehow missed that we do not use 4 bit colors anymore.
I have tried to run application without a desktop in virtual console of Linux.

![photo_9_2024-12-23_05-13-15](https://github.com/user-attachments/assets/cf5a04d8-bf41-48bb-96f3-edb06eceef86)
![photo_8_2024-12-23_05-13-15](https://github.com/user-attachments/assets/59fffd64-179d-485c-84b0-71a92fe3a537)
![photo_7_2024-12-23_05-13-15](https://github.com/user-attachments/assets/b9d92f7f-dc55-409f-a3e4-108ee8d841c0)
![photo_6_2024-12-23_05-13-15](https://github.com/user-attachments/assets/27e2252c-f793-4d13-99c2-a7aa6c765d27)
![photo_5_2024-12-23_05-13-15](https://github.com/user-attachments/assets/854ee684-0404-49ed-af75-bc9ea8e19e44)
![photo_4_2024-12-23_05-13-15](https://github.com/user-attachments/assets/3d99b4e7-26ba-4ae8-9b54-7496ef50b676)
![photo_3_2024-12-23_05-13-15](https://github.com/user-attachments/assets/9e991b39-3afa-4f4c-ad15-9d71000ecfad)
![photo_2_2024-12-23_05-13-15](https://github.com/user-attachments/assets/a8d7e117-e045-4a12-ac61-43727f41de98)
![photo_1_2024-12-23_05-13-15](https://github.com/user-attachments/assets/d8069f97-7087-4986-8fdf-34e86f8338d8)

It looks like that was the case when 256 colors not supported.
But now I do not understand how we should deal with this case.

**Reply by jinek** [reply]
Yeah. It looks like it's simulating CGA output https://en.wikipedia.org/wiki/Linux_console

**Reply by jinek** [reply]
Some info. ChatGPT outputs this:

```
Command:

echo $TERM

Interpretation:

    linux: Typically refers to Linux virtual consoles (TTYs), which usually support 16 colors.
    xterm-256color: Indicates support for 256 colors.
    screen-256color, tmux-256color: Also signify 256-color support within terminal multiplexers.
    xterm-direct, xterm-color: Varying levels of color support, often 16 to 256 colors.
```

**Reply by jinek** [reply]
![IMG_20241223_172854_155.jpg](https://github.com/user-attachments/assets/05ace470-7a23-42b7-8027-9607f15049a1)

![IMG_20241223_172903_070.jpg](https://github.com/user-attachments/assets/c67a31c7-f640-4205-bd61-977ee6a62357)

**Reply by jinek** [reply]
With all this now the question is, if we detected 16 colors only, should we
1) properly project 8bit colors to 4 bits and adjust "shading" operation to produce a 4 bit color if source was 4 bits
Or
2) should we bring back ConsoleColor to work in parallel to 8bit?

@tomlm or do we have any other options?

With 2nd option it will not be possible to use fluent theme in 16colors mode. While with second option i can not predict how its gonna look like.
Also anyway in this terminals stuff like edge border, understrike etc does not work.

**Reply by tomlm** [reply]
Btw I  think the 256 color escape sequence actual has an escape sequence to define with each of the 256 color are as a table. Then when you output color 99 you get the 99 color from the table

**Reply by jinek** [reply]
What does it give to us?

**Reply by tomlm** [reply]
In theory you could automatically gather the colors and only have to interpolate after you have used all 256 slots
